File: MsgSendDemo/MsgSendDemo/ViewController.m
//
//  ViewController.m
//  MsgSendDemo
//
//  Created by AugustRush on 6/17/16.
//  Copyright © 2016 August. All rights reserved.
//

#import "ViewController.h"
#import <objc/runtime.h>
#import <objc/message.h>

#define TYPE(...)

#define MsgSend1(caller,method,arg) (((id(*)(__typeof__(caller),SEL,__typeof__(arg)))objc_msgSend)(caller,@selector(method),arg))

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
//    id r = (((id(*)(id,SEL,id))objc_msgSend)(self.view,@selector(setBackgroundColor:),[UIColor purpleColor]));
    id r = MsgSend1(self.view, setBackgroundColor:, [UIColor redColor]);
    NSLog(@"result is %@",r);
    
    NSLog(@"UIView alloc is %@",[UIView alloc]);
    
    class_addProtocol([UIView class], @protocol(UIViewExport));
    
    JSContext *context = [[JSContext alloc] init];
    context[@"UIView"] = [UIView class];
    
    [context setExceptionHandler:^(JSContext *context, JSValue *value) {
        NSLog(@"context error is %@",context.exception);
    }];
    UIView *view = [[context evaluateScript:@"UIView.new()"] toObject];
    UIView *view1 = [[context evaluateScript:@"UIView.alloc().init()"] toObject];
    NSLog(@"jsView is %@,\n view1 is %@",view,view1);
    
    NSString *file = [[NSBundle mainBundle] pathForResource:@"Test" ofType:@"js"];
    NSString *mustacheJSString = [NSString stringWithContentsOfFile:file encoding:NSUTF8StringEncoding error:nil];
    JSValue *scriptValue = [context evaluateScript:mustacheJSString];
    NSObject *object = [context[@"view"] toObject];
    NSLog(@"object is %@",object);
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: JSDemo/JSDemo/ViewController.swift
//
//  ViewController.swift
//  JSDemo
//
//  Created by AugustRush on 5/7/16.
//  Copyright © 2016 August. All rights reserved.
//

import UIKit
import JavaScriptCore

class ViewController: UIViewController {

    let context = JSContext()
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        print(self.showAlert)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
        
        let simplifyString: @convention(block) String -> String = { input in
//            let mutableString = NSMutableString(string: input) as CFMutableStringRef
//            CFStringTransform(mutableString, nil, kCFStringTransformToLatin, false)
//            CFStringTransform(mutableString, nil, kCFStringTransformStripCombiningMarks, false)
//            return mutableString as String
            self.showAlert()
            return ""
        }
        context.setObject(unsafeBitCast(simplifyString, AnyObject.self), forKeyedSubscript: "simplifyString")
        
        print(context.evaluateScript("simplifyString('안녕하새요!')"))
        // annyeonghasaeyo!
    }
    
    dynamic func showAlert() {
        let alert = UIAlertController(title: "title",message: "message text",preferredStyle: .Alert);
        self.showViewController(alert, sender: alert)
    }

}

